### PR TITLE
wallet: Skip transactions with no shielded data in CWallet::SetBestChain()

### LINKED ISSUE #3495: Failed atomic write of CWalletTx is stopping it from storing the last block scanned
```2018-08-29 03:11:07 SetBestChain(): Failed to write CWalletTx, aborting atomic write```

when we restarted with the new version

```2018-08-28 23:45:02 Rescanning last 43728 blocks (from block 340372)...```
```2018-08-28 21:45:27 Verifying last 288 blocks at level 3
2018-08-28 21:45:30 No coin database inconsistencies in last 289 blocks (2284 transactions)
2018-08-28 21:45:30  block index           16798ms
2018-08-28 21:45:30 init message: Loading wallet...
2018-08-28 23:45:02 nFileVersion = 1010150
2018-08-28 23:45:02 Keys: 146155 plaintext, 0 encrypted, 146155 w/ metadata, 146155 total
2018-08-28 23:45:02 ZKeys: 2 plaintext, 0 encrypted, 2 w/metadata, 2 total
2018-08-28 23:45:02  wallet              7171409ms```

it took almost 2 hours to load the wallet

```2018-08-29 03:02:35  rescan             11852755ms
2018-08-29 03:11:07 SetBestChain(): Failed to write CWalletTx, aborting atomic write```

3 hours to rescan the entire thing
and i'm almost positive that the failed write to cwallettx is stopping it from storing the last block scanned so in restart, we will suffer the 3 hour rescan again.

 ```aborting debug.log
2018-08-28 05:37:34 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 06:38:19 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 07:42:16 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 08:44:18 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 09:47:20 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 10:49:52 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 11:52:08 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 12:54:25 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 13:55:35 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 14:58:56 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 16:00:08 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 17:00:07 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 18:01:38 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 19:06:48 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-28 20:14:58 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 03:11:07 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 03:20:25 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 04:20:48 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 05:21:16 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 06:23:49 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 07:23:53 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 08:26:12 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 09:35:21 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 10:35:49 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 11:36:46 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 12:42:10 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 13:50:57 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 14:52:12 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 15:55:54 SetBestChain(): Failed to write CWalletTx, aborting atomic write
2018-08-29 16:56:11 SetBestChain(): Failed to write CWalletTx, aborting atomic write```


this basically happens ever hour

### PR DESCRIPTION
**Pull request by str4d**
Co-authored-by: Daira Hopwood <daira@jacaranda.org>

Closes #3495.

### CODE REVIEW

File: src/wallet/gtest/test_wallet.cpp [hunk 1]
```diff
@@ -1627,6 +1627,63 @@ TEST(WalletTests, WriteWitnessCache) {
...
+    auto sk = libzcash::SproutSpendingKey::random();
+    wallet.AddSproutSpendingKey(sk);
+
+    // Generate a transparent transaction that is ours
+    CMutableTransaction t;
+    t.vout.resize(1);
+    t.vout[0].nValue = 90*CENT;
+    t.vout[0].scriptPubKey = scriptPubKey;
+    CWalletTx wtxTransparent {nullptr, t};
+    wallet.AddToWallet(wtxTransparent, true, NULL);
```
Comment (daira) [inline src/wallet/gtest/test_wallet.cpp]: s/NULL/nullptr/g

File: src/wallet/gtest/test_wallet.cpp [hunk 2]
```diff
@@ -1627,6 +1627,63 @@ TEST(WalletTests, WriteWitnessCache) {
...
+
+    // Generate a Sprout transaction that only involves our transparent address
+    auto sk2 = libzcash::SproutSpendingKey::random();
+    auto wtxInput = GetValidReceive(sk2, 10, true);
+    auto note = GetNote(sk2, wtxInput, 0, 0);
+    auto wtxTmp = GetValidSpend(sk2, note, 5);
+    CMutableTransaction mtx {wtxTmp};
+    mtx.vout[0].scriptPubKey = scriptPubKey;
+    CWalletTx wtxSproutTransparent {NULL, mtx};
+    wallet.AddToWallet(wtxSproutTransparent, true, NULL);
```
Comment (daira) [inline src/wallet/gtest/test_wallet.cpp]: Do we want to also include a Sapling tx?

**Comment by str4d** [comment]
Failed on the kitchensink builder due to low disk space. The other supported builders all passed.

**Comment by Eirik0** [comment]
@zkbot r+

**Comment by Eirik0** [comment]
@zkbot r+

**Comment by Eirik0** [comment]
Looks like intermittent RPC test failures.

**Comment by Eirik0** [comment]
@zkbot retry

**Comment by str4d** [comment]
Another transient failure.

@zkbot retry

**Comment by daira** [comment]
False-positive test failures are really getting in the way of development at this point :-(

**Comment by str4d** [comment]
Another transient failure (in the same RPC test, now across three different
builders - its timing tolerances are probably too low for AWS).

@zkbot retry

**Comment by str4d** [comment]
Okay, this RPC test is failing too often for it to be purely AWS-related.

**Comment by garethtdavies** [comment]
It's this issue isn't it? https://github.com/zcash/zcash/issues/3718

**Comment by Eirik0** [comment]
@zkbot retry

**Comment by Eirik0** [comment]
The failure looks like it has changed to being a timeout. Giving it one more retry and then we may need to look deeper.

**Comment by Eirik0** [comment]
@zkbot retry

**Pull request merged**